## Chapter 3

## Problem 1

Show that the F statistic (3.13) for dropping a single coefficient
from a model is equal to the square of the corresponding z-score (3.12).

$p_1-p_0=1$, $RSS = (Y-X\beta)^T(Y-X\beta)$, 

$\frac{(RSS_0-RSS_1)}{RSS_1/(N-p_1-1)} = \beta_j^2/(\sigma^2(X^TX)_j)$

## Problem 2

Given data on two variables X and Y , consider fitting a cubic
polynomial regression model $f(X) =
\sum^3_{j=0} \beta_jX_j$. In addition to plotting
the fitted curve, you would like a 95% confidence band about the curve.
Consider the following two approaches:
1. At each point x_0, form a 95% confidence interval for the linear function
$a^T\beta =\sum^3_{j=0} \beta_jx_j$

2. Form a 95% confidence set for β as in (3.15), which in turn generates
confidence intervals for f(x0).

How do these approaches differ? Which band is likely to be wider? Conduct
a small simulation experiment to compare the two methods.


The first approach consider percentile for normal distribution for each $\beta$. The second approach we consider percentile for chi distribution for set of $\beta$s. I would expect second band to be wider all beta contribute at the same time.

In [32]:
X=rnorm(20)
Y = 2*X+X^2+0.6*X^3+rnorm(20)
df = data.frame(X, Y)
head(df)

X           Y         
1  1.99676405 12.5180369
2  0.14887276  1.1051888
3  0.02975812 -0.9762164
4 -0.10232201 -2.0374831
5  1.30950677  5.1682533
6 -1.64164290 -4.0566385

In [33]:
m = lm(Y~X+I(X^2)+I(X^3), data=df)
confint(m)

2.5 %      97.5 %     
(Intercept) -1.1396592 -0.01312149
X            0.9334383  3.42163117
I(X^2)       0.5351249  1.51890494
I(X^3)       0.1749300  0.97235379

In [34]:
summary(m)


Call:
lm(formula = Y ~ X + I(X^2) + I(X^3), data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-1.6461 -0.4081 -0.1458  0.4246  2.1160 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -0.5764     0.2657  -2.169 0.045470 *  
X             2.1775     0.5869   3.710 0.001900 ** 
I(X^2)        1.0270     0.2320   4.426 0.000424 ***
I(X^3)        0.5736     0.1881   3.050 0.007639 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.9701 on 16 degrees of freedom
Multiple R-squared:  0.978,	Adjusted R-squared:  0.9739 
F-statistic: 237.1 on 3 and 16 DF,  p-value: 1.812e-13


In [35]:
m = lm(Y~poly(X,3), data=df)
confint(m)

2.5 %      97.5 %   
(Intercept)  1.8172327  2.736903
poly(X, 3)1 22.5039022 26.616793
poly(X, 3)2  5.5220314  9.634923
poly(X, 3)3  0.9022406  5.015132

In [36]:
summary(m)


Call:
lm(formula = Y ~ poly(X, 3), data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-1.6461 -0.4081 -0.1458  0.4246  2.1160 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   2.2771     0.2169  10.498 1.39e-08 ***
poly(X, 3)1  24.5603     0.9701  25.318 2.46e-14 ***
poly(X, 3)2   7.5785     0.9701   7.812 7.53e-07 ***
poly(X, 3)3   2.9587     0.9701   3.050  0.00764 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.9701 on 16 degrees of freedom
Multiple R-squared:  0.978,	Adjusted R-squared:  0.9739 
F-statistic: 237.1 on 3 and 16 DF,  p-value: 1.812e-13


At least if to consider poly function the confidence intervals are wider for the same error

## Problem 3

Gauss–Markov theorem:
(a) Prove the Gauss–Markov theorem: the least squares estimate of a
parameter $a^T$β has variance no bigger than that of any other linear
unbiased estimate of $a^T$β (Section 3.2.2).

Let take estimater which differs from least square estimator by bector $b^T$. So that

$E(c^T Y) = E(a^T\beta+b^TY) = E(a^T\beta) + E(b^tX\beta)$

because we have unbias estimator 

$=a^T\beta$ Eq.(3.18)

$b^TX = 0$, taking this into account to remove cross products

$(c^Ty) = c^TVar(y)c = \sigma^2 c^Tc = Var(a^T\beta - b^TY)(a^T \beta - b^TY)^T = \sigma^2 (a^Ta/(X^TX)+b^Tb) > Var(a^T\beta)$

(b) The matrix inequality B =< A holds if A − B is positive semidefinite.
Show that if $\hat{V}$ is the variance-covariance matrix of the least squares
estimate of β and $V$ is the variance-covariance matrix of any other
linear unbiased estimate, then $\hat{V} =< V$.

Positive semidefinite $X^T(A-B)X >= 0$ for any real X. $\hat{V} = E(a^T\beta a\beta^T)-E(a^T\beta)E(a\beta^T)$, $V = E(c^T\beta c\beta^T)-E(c^T\beta)E(c\beta^T)$ from part (a) = $\hat{V}+E(b^Tb)>=\hat{V}$

## Problem 4

Show how the vector of least squares coefficients can be obtained
from a single pass of the Gram–Schmidt procedure (Algorithm 3.1). Represent
your solution in terms of the QR decomposition of X.

By definition $\beta=X^TY/(XX^{T})$, $x_j =z_j + \sum_{k=1}^{j-1} \gamma_{jk}z_k $ or $X=Z\Gamma$

QR dicomposition: $X=QR$, where Q is orthogonal matrix, R is upper triangular matrix, Z=QD, $\Gamma=D^{-1}R$, D is diagonal matrix.

$\beta = Q^TR^TY/(QRQ^{T}R^{T})=Q^TY/R$

Using that R is upper triangular matrix (see also wiki QR decomposition)
$r_{pp}\beta_{p} = <q_{p}, y>$

$\beta_{p} = <z_p, y>/||z_{p}||^2$

Substituting back for equation for $\beta_j$,

$r_{p-1, p-1}\beta_{p-1} +r_{p-1, p}\beta_p=<q_p, y>$  

$\beta_{p-1} +\gamma_{p-1,p}\beta_p=<z_{p-1}, y>/||z_{p-1}||^2$  

## Problem 5

Consider the ridge regression problem (3.41). Show that this problem
is equivalent to the problem

$\hat{\beta^c}= argmin_{\beta_c}{(\sum_{i=1}^{N}(y_i-\beta_0^c-\sum_{j=1}^p(x_{ij}-\bar{x}_j)\beta^c_j)^2+\lambda\sum_{j=1}^p\beta_j^{c2})}$

Give the correspondence between βc and the original β in (3.41). Characterize
the solution to this modified criterion. Show that a similar result
holds for the lasso.


We can write $\hat{\beta^c}= argmin_{\beta_c}{(\sum_{i=1}^{N}(y_i-(\beta_0^c-\sum_{j=1}^p\bar{x}_{j})-\sum_{j=1}^px_{ij}\beta^c_j)^2+\lambda\sum_{j=1}^p\beta_j^{c2})}$

$\beta_0 = \beta_0^c-\sum_{j=1}^p\bar{x}_{j}$, $\beta_j=\beta_j^c$ $j>0$

It does not depend on regulization term if it does not contain $\beta_0$

## Problem 6

Show that the ridge regression estimate is the mean (and mode)
of the posterior distribution, under a Gaussian prior $\beta ~ N(0, \tau I)$, and
Gaussian sampling model $y ~ N(X\beta, \sigma^2I)$. Find the relationship between
the regularization parameter $\lambda$ in the ridge formula, and the variances $\tau$
and $\sigma^2$.

$\hat{\beta}_R = X^Ty(X^TX+\lambda I)^{-1}$

From Gaussian distribution:

$\hat{\beta}_R = argmax_{\beta}(L(y|\beta)\pi(\beta))$, where L is Gaussian likelyhood, $\pi(\beta)$ is probability, $p(\beta|y)=p(y|\beta)\pi(\beta)$.
Taking log from argument

$\hat{\beta}_R = argmax_{\beta}(-(y-X\beta)^T(y-X\beta)/\sigma^2-\beta^T\beta/\tau^2)=argmin_{\beta}((\beta^TX^TX\beta-\beta^TX^Ty-y^TX\beta)/\sigma^2+\beta^T\beta/\tau^2)$ 

If gradient by $\beta$ in argmin is 0 we have

$\hat{\beta}_R = X^Ty(X^TX + \sigma^2/\tau^2  I)^{-1}$

## Problem 7

## Problem 8

Consider the QR decomposition of the uncentered N × (p + 1)
matrix X (whose first column is all ones), and the SVD of the N × p
centered matrix $\hat{X}$. Show that Q2 and U span the same subspace, where
Q2 is the sub-matrix of Q with the first column removed. Under what
circumstances will they be the same, up to sign flips?

Q is orthoganal matrix the columns of Q are $q_1, .., q_{p+1}$.  By QR decomposition $span(q_i,... q_j) = span(x_i, ..,x_j), 1<=j<=p+1$ or for Q2 if we shift indexes by -1  $span(q_i,... q_j) = span(x_i, ..,x_j), 1<=j<=p$ (1)

We can write $\hat{x}_i = \sum_{j<i} a_jq_j$, so $span(q_1, .., q_i)=span(x_1,.., x_i)$ (2)
For SVD $span(u_1,.., u_p) = span(\hat{x}_1,.., \hat{x}_p)$ (3). Therefore from (1) -(3) $span(q_i,... q_j) = span(u_1,.., u_p)$, Q2 and U span the same subspace.

If $sign(\hat{x_i})=sign(x_i)$(X is without original first column) for any $i$ they must be the same up to sign flip.

## Problem 9

Forward stepwise regression. Suppose we have the QR decomposition
for the N×q matrix X1 in a multiple regression problem with response
y, and we have an additional p−q predictors in the matrix X2. Denote the
current residual by r. We wish to establish which one of these additional
variables will reduce the residual-sum-of squares the most when included
with those in X1. Describe an efficient procedure for doing this.

Add $X_{21}$ to matrix $X_{1}$, do QR decomposition, find $\beta$ and new residual r. Compare it to initial r if r get smaller save $X_{21}$ and corresponding sum. Try the same on next, compare with the current saved smallest r. Do it for all predictors in $X_2$. This way find which predictor has smallest r.

## Problem 10

Backward stepwise regression. Suppose we have the multiple regression
fit of y on X, along with the standard errors and Z-scores as in
Table 3.2. We wish to establish which variable, when dropped, will increase
the residual sum-of-squares the least. How would you do this?

Consider most of p-1 models removing 1 feature. Choose model with smallest  residual sum-of-squares. I would start from highest absolute value of z score, but not rely too much on it. If the residual sum-of-squares really started to increase more for smaller Z score and Z scores distribution does not change much in new models, then it may be Ok not to check features with smallest  scores.

## Problem 11

Show that the solution to the multivariate linear regression problem (3.40) is given by (3.39). What happens if the covariance matrices $\sum_i$
are different for each observation?

If we have the same $\epsilon_k$ for all observations it does not participate in $\sum_i$, minimizing $RSS(B, \sum)$ we come
again to equation $X^T\sum^{-1}(Y-X\hat{B}) = 0$ in case if $\sum$ is independent from $i$, $\hat{B}=X^TY/(X^TX)$. If $\sum_ik$
are deferent for different i, we have like different weights in equation and we must leave $\sum$ in it  $\hat{B}=X^T\sum^{-1}Y/(X^TX)$

## Problem 12

Show that the ridge regression estimates can be obtained by
ordinary least squares regression on an augmented data set. We augment
the centered matrix X with p additional rows $\sqrt{\lambda}I$, and augment y with p
zeros. By introducing artificial data having response value zero, the fitting
procedure is forced to shrink the coefficients toward zero. This is related to
the idea of hints due to Abu-Mostafa (1995), where model constraints are
implemented by adding artificial data examples that satisfy them.

Loss function is $\sum_{j=1}^{N+p}\sum_{k=0}^p(y_{j}-x_{jk}\beta_k)^2=\sum_{j=1}^{N}\sum_{k=0}^p(y_{j}-x_{jk}\beta_k)^2+\sum_{j=N+1}^{N+p}\lambda(\sum_{k=0}^p I_{jk}\beta_k)^2 =\sum_{j=1}^{N}\sum_{k=0}^p(y_{j}-x_{jk}\beta_k)^2+\sum_{k=0}^p\lambda\beta_k^2$

## Problem 13

Derive the expression (3.62), and show that $\hat{\beta}_{pcr}(p) = \hat{\beta}_{ls}$.

$z_m = Xv_m$, substituting in (3.61) we get $\hat{y}_{pcr} =\bar{y}{\bar{1}}+ X\sum_{m=1}^M\hat{\theta}_mv_m$, $\hat{\beta}_{pcr} = \sum_{m=1}^M\hat{\theta}_mv_m = \hat{\beta}_{ls}$, because $\hat{y}_{ls}=\bar{y}{\bar{1}}+X\hat{\beta}_{ls}$

## Problem 14

Show that in the orthogonal case, PLS stops after m = 1 steps,
because subsequent $\hat{\phi}_{mj}$ in step 2 in Algorithm 3.3 are zero.

Consider $x_j^{(m)}$, equation in part 2(d). In orthogonal case in part 2 (d) of Algorithm 3.3 we obtain that in $<z_m, x_j^{(m-1)}>=\hat{\phi}_{mj}x_j^{(m-1)2}$, m=2, we have only this component. Then $<x_j^{(m)},y> = <x_j^{(m-1)},y>-(\hat{\phi}_{mj}x_j^{(m-1)2}<z_m, y>)/\sum_{j=1}^p{\hat{\phi}_{mj}^2x_j^{(m-1)2}}=\hat{\phi}_{mj}-\hat{\phi}_{mj}=0$, if m>1, $z_m = \sum_{j=1}^p \phi_{mj}x^{(m-1)}_j$.

## Problem 15

Verify expression (3.64), and hence show that the partial least
squares directions are a compromise between the ordinary regression coefficient
and the principal component directions.

Algorithm 3.3 is maximization of covariance matrix $cov_{||\alpha=1||}(Y, \alpha X)^2 = corr(Y, \alpha X)^2Var(Y) Var(\alpha X)$

$cov_{||\alpha=1||}(Y, \alpha X)^2 = \alpha^TX^TYY^TX\alpha$

Condition of maximum eigen value and eigen vector pair of this symmetic matrix $X^TYY^TX$  is 3.64

In Algorithm 3.3 $\phi_{mj}$ are products $<x_{j}, y>$ at the end we come to 3.64.
